collect_app/src/main/java/org/odk/collect/android/logic/FormController.java

```diff
@@ -396,6 +396,20 @@ public String getAppearanceAttr(@NonNull FormIndex index) {
         return element.getAppearanceAttr();
     }
 
+    /**
+     * Returns true if the question at the given FormIndex uses the search() appearance/function
+     * of "fast itemset" feature.
+     *
+     * Precondition: there is a question at the given FormIndex.
+     */
+    public boolean usesDatabaseExternalDataFeature(@NonNull FormIndex index) {
+        String queryAttribute = getFormDef().getChild(index).getAdditionalAttribute(null, "query");
+        String appearanceAttribute = getAppearanceAttr(index);
+
+        return appearanceAttribute != null && ExternalDataUtil.SEARCH_FUNCTION_REGEX.matcher(appearanceAttribute).find()
+                || queryAttribute != null && queryAttribute.length() > 0;
```

Comment:
It would be good to factor out a method that checks if it is `ItemsetWIdget` because we check it in other places as well not only here and such a method would be more meaningful instead of using `queryAttribute != null && queryAttribute.length() > 0`

Reply:
That sounds good for the other contexts but in this particular case we want to know if a question at a `FormIndex` uses the feature and we don't have a widget to work with. `FormController` should not know anything about `QuestionWidget` and same with the `updateFieldList` method. I agree that statement is a bit opaque but that's why it's in a method with a comment rather than inline. You'd prefer looking up the widget from the `FormIndex` first? Where would you put the method you're describing?

Reply:
I was thinking about the new [WidgetAppearanceUtils](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/utilities/WidgetAppearanceUtils.java) class (at least to factor out a method that checks if it's `ItemsetWidget`) but it's not super important so we can leave it as is, for now, since there is a comment.